File: consys-core/src/main/scala/de/tuda/stg/consys/objects/Ref.scala
package de.tuda.stg.consys.objects

/**
	* Created on 28.02.19.
	*
	* @author Mirko Köhler
	*/
trait Ref[Addr, T <: AnyRef] extends Serializable {

	val addr : Addr
	val consistencyLevel : ConsistencyLevel

	def deref : ReplicatedObject[Addr, T]

	def isAvailable : Boolean

	def await() : Unit

	def delete() : Unit


	/* shortcut for Java implementation */
	final def ref : T = deref.ref
}

File: consys-japi/src/main/java/de/tuda/stg/consys/japi/impl/cassandra/Ref.java
package de.tuda.stg.consys.japi.impl.cassandra;

/**
 * Created on 27.01.20.
 *
 * @author Mirko Köhler
 */
public interface Ref<T> {

	default T ref() {
		throw new UnsupportedOperationException("Use the consys-compiler plugin to resolve calls to ref().");
	}

	<R> R getField(String fieldName);

	<R> R setField(String fieldName, R value);

	<R> R invoke(String methodName, Object... args);
}
